### ISSUE TITLE
Bootstrapping a new project

### ISSUE THREAD

**Issue by asheshambasta** [Gabriel439/haskell-nix#86]
Following this guide, I've been trying to bootstrap a nix based project, but I'm facing some issues.
For starters, here's the minimal project: https://github.com/asheshambasta/avahi-dbus 
I've created a cabal file without any version bounds, declared all the dependencies and wrote the appropriate nix expressions. 

Given that the package `dbus` has foreign dependencies, I've also added `zlib` as its dependency. 

---

Next, I'm firing up a nix-shell with `nix-shell release.nix` which seems to pull in dependencies, after which I'm attempting to start a repl with `cabal repl`.

This step fails with:
```
<command line>: can't load .so/.DLL for: libz.so (libz.so: cannot open shared object file: No such file or directory) 
```

Which has me confused. `zlib` is a dependency that should be present on the path but it seems like Cabal cannot find it. I also came across this https://github.com/NixOS/nixpkgs/issues/56860 which seems related but I'm not sure. I'm assuming the issue I'm running into is so common that more people would've been talking about it: needless to say that the linked issue has been closed. 

Also, here's my nix-info: 

```
system: "x86_64-linux", multi-user?: yes, version: nix-env (Nix) 2.3.3, channels(root): "nixos-19.09.2335.ace3bb36fed, oldpkgs-19.03.173676.d1dff0bcd9f", nixpkgs: /nix/var/nix/profiles/per-user/root/channels/nixos
```

--- 

What can I do to circumvent this? Or is this a known cabal/ghci issue?

**Comment by Gabriel439** [comment]
@asheshambasta: The short answer is that you can fix the problem by running:

```shell
$ nix-shell --attr avahi-dbus.env release.nix
```

The issue here is that running `nix-shell` directly on the derivation for the Haskell derivation does not behave well (unlike almost all other Nix packages).  Instead, every Haskell derivation has an `env` attribute that you have to use `nix-shell` on instead.  See the beginning of this section for more details:

https://github.com/Gabriel439/haskell-nix/tree/master/project0#building-with-cabal

**Comment by asheshambasta** [comment]
@Gabriel439 thanks a lot for clarifying this. This indeed fixes the issue. 
Is there a more detailed resource on why this is with Haskell packages?

I'm closing this issue since this seems fixed for me. 
Apologies for a question-issue.

**Comment by Gabriel439** [comment]
@asheshambasta: There isn't an explanation of why this is the case.  That idiom is taken from the [Haskell section of the Nixpkgs manual](https://nixos.org/nixpkgs/manual/#haskell), but it doesn't explain the rationale behind it.

**Comment by asheshambasta** [comment]
@Gabriel439 Thanks for the link. I'll read it and respond back.

I'm not sure where else to ask, but following: https://github.com/Gabriel439/haskell-nix/blob/master/project2/README.md, I've added the zlib dependency in the avahi-dbus.nix, but builds still fail with the `extra-libraries` option in the cabal file. 
I've added a commit that demonstrates this: https://github.com/asheshambasta/avahi-dbus/commit/f8a342300448dc3f2a9a9bab4a7bbeea73e091f6
